### PR TITLE
Fix redundant iteration while splitting lines

### LINKED ISSUE #13551: Bug: PowerShell cannot handle a file with long non-ascii file name
## Steps to reproduce
Create a directory named "Test"
In the directory, create a file named "测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试测试.txt"
```powershell
cd Test
ls
```

## Expected behavior
Files listed normally.

## Actual behavior

```none
out-lineoutput: startIndex cannot be larger than length of string. (Parameter 'startIndex')
```

## Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.3
PSEdition                      Core
GitCommitId                    7.0.3
OS                             Microsoft Windows 10.0.20201
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by hez2010**
# PR Summary

Fix redundant iteration while splitting lines to prevent assert failure.

## PR Context

We split a line to fit in a windows width and if on last iteration we did get zero length tail we did next iteration but we should stop.

Borrowed from @iSazonov, see https://github.com/PowerShell/PowerShell/issues/13551#issuecomment-766722563

Fixes #13551

Todo: 

- ~[ ] Add tests~: can only repro in a specified range of width of window while rendering table, I don't know how to test it automatically without user interaction.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Review by iSazonov — APPROVED** [review]

File: src/System.Management.Automation/FormatAndOutput/common/ComplexWriter.cs [hunk 1]
```diff
@@ -416,10 +416,12 @@ private static StringCollection GenerateLinesWithoutWordWrap(DisplayCells displa
 
             for (int k = 0; k < lines.Length; k++)
             {
-                if (lines[k] == null || displayCells.Length(lines[k]) <= firstLineLen)
+                var currentLine = lines[k];
```
Comment (xtqqczze) [inline src/System.Management.Automation/FormatAndOutput/common/ComplexWriter.cs:419]:
```suggestion
                string currentLine = lines[k];
```

We should respect EditorConfig setting: `csharp_style_var_for_built_in_types = false`.

File: src/System.Management.Automation/FormatAndOutput/out-console/ConsoleLineOutput.cs [hunk 2]
```diff
@@ -87,8 +87,15 @@ internal DisplayCellsPSHost(PSHostRawUserInterface rawUserInterface)
 
         internal override int Length(string str, int offset)
         {
```
Comment (xtqqczze) [inline src/System.Management.Automation/FormatAndOutput/out-console/ConsoleLineOutput.cs:89]: Are the changes to this method necessary to fix the issue?

File: src/System.Management.Automation/FormatAndOutput/out-console/ConsoleLineOutput.cs [hunk 3]
```diff
@@ -87,8 +87,15 @@ internal DisplayCellsPSHost(PSHostRawUserInterface rawUserInterface)
...
-            Dbg.Assert(offset >= 0, "offset >= 0");
-            Dbg.Assert(string.IsNullOrEmpty(str) || (offset < str.Length), "offset < str.Length");
+            if (string.IsNullOrEmpty(str))
+            {
+                return 0;
+            }
+
+            if (offset < 0 || offset >= str.Length)
+            {
+                throw PSTraceSource.NewArgumentException(nameof(offset));
```
Comment (xtqqczze) [inline src/System.Management.Automation/FormatAndOutput/out-console/ConsoleLineOutput.cs:97]: In particular this exception.

**Comment by hez2010** [comment]
@iSazonov can I push to your branch? or I have to create another PR from my fork and close this one.

File: src/System.Management.Automation/FormatAndOutput/out-console/ConsoleLineOutput.cs (hunk 2 above)
Comment (hez2010) [reply src/System.Management.Automation/FormatAndOutput/out-console/ConsoleLineOutput.cs:89]: No. But it's a refactor to `Length` to reduce ops while `str` is empty.

**Comment by iSazonov** [comment]
> @iSazonov can I push to your branch? or I have to create another PR from my fork and close this one.

It is better to use your fork if you want to add something. Please avoid in-depth refactoring.

**Comment by hez2010** [comment]
I'm closing this PR and create another one from my folk later.

**Pull request closed without merging**